src/machine/machine_rp2040_usb.go

```diff
@@ -0,0 +1,309 @@
+
+	//// Mux the controller to the onboard usb phy
+	rp.USBCTRL_REGS.USB_MUXING.Set(rp.USBCTRL_REGS_USB_MUXING_TO_PHY | rp.USBCTRL_REGS_USB_MUXING_SOFTCON)
+
+	//// Force VBUS detect so the device thinks it is plugged into a host
+	rp.USBCTRL_REGS.USB_PWR.Set(rp.USBCTRL_REGS_USB_PWR_VBUS_DETECT | rp.USBCTRL_REGS_USB_PWR_VBUS_DETECT_OVERRIDE_EN)
+
+	//// Enable the USB controller in device mode.
+	rp.USBCTRL_REGS.MAIN_CTRL.Set(rp.USBCTRL_REGS_MAIN_CTRL_CONTROLLER_EN)
+
+	//// Enable an interrupt per EP0 transaction
+	rp.USBCTRL_REGS.SIE_CTRL.Set(rp.USBCTRL_REGS_SIE_CTRL_EP0_INT_1BUF)
+
+	//// Enable interrupts for when a buffer is done, when the bus is reset,
+	//// and when a setup packet is received
+	rp.USBCTRL_REGS.INTE.Set(rp.USBCTRL_REGS_INTE_BUFF_STATUS |
+		rp.USBCTRL_REGS_INTE_BUS_RESET |
+		rp.USBCTRL_REGS_INTE_SETUP_REQ)
+
+	//// Present full speed device by enabling pull up on DP
+	rp.USBCTRL_REGS.SIE_CTRL.SetBits(rp.USBCTRL_REGS_SIE_CTRL_PULLUP_EN)
+}
+
+func handleUSBIRQ(intr interrupt.Interrupt) {
+	status := rp.USBCTRL_REGS.INTS.Get()
+
+	// Setup packet received
+	if (status & rp.USBCTRL_REGS_INTS_SETUP_REQ) > 0 {
+		rp.USBCTRL_REGS.SIE_STATUS.Set(rp.USBCTRL_REGS_SIE_STATUS_SETUP_REC)
+		setup := newUSBSetup(usbDPSRAM.setupBytes())
+
+		ok := false
+		if (setup.BmRequestType & usb_REQUEST_TYPE) == usb_REQUEST_STANDARD {
+			// Standard Requests
+			ok = handleStandardSetup(setup)
+		} else {
+			// Class Interface Requests
+			if setup.WIndex < uint16(len(usbSetupHandler)) && usbSetupHandler[setup.WIndex] != nil {
+				ok = usbSetupHandler[setup.WIndex](setup)
+			}
+		}
+
+		if ok {
+			// set Bank1 ready?
+		} else {
+			// Stall endpoint?
+		}
+
+	}
+
+	// Buffer status, one or more buffers have completed
+	if (status & rp.USBCTRL_REGS_INTS_BUFF_STATUS) > 0 {
+		if sendOnEP0DATADONE.offset > 0 {
+			ep := uint32(0)
+			data := sendOnEP0DATADONE.data
+			count := len(data) - sendOnEP0DATADONE.offset
+			if ep == 0 && count > usbEndpointPacketSize {
+				count = usbEndpointPacketSize
+			}
+
+			sendViaEPIn(ep, data[sendOnEP0DATADONE.offset:], count)
+			sendOnEP0DATADONE.offset += count
+			if sendOnEP0DATADONE.offset == len(data) {
+				sendOnEP0DATADONE.offset = 0
+			}
+		}
+
+		s2 := rp.USBCTRL_REGS.BUFF_STATUS.Get()
+
+		// OUT (PC -> rp2040)
+		for i := 0; i < 16; i++ {
+			if s2&(1<<(i*2+1)) > 0 {
+				buf := handleEndpointRx(uint32(i))
+				if usbRxHandler[i] != nil {
+					usbRxHandler[i](buf)
+				}
+			}
+		}
+
+		// IN (rp2040 -> PC)
+		for i := 0; i < 16; i++ {
+			if s2&(1<<(i*2)) > 0 {
+				if usbTxHandler[i] != nil {
+					usbTxHandler[i]()
+				}
+			}
+		}
+
+		rp.USBCTRL_REGS.BUFF_STATUS.Set(0xFFFFFFFF)
+	}
+
+	// Bus is reset
+	if (status & rp.USBCTRL_REGS_INTS_BUS_RESET) > 0 {
+		rp.USBCTRL_REGS.SIE_STATUS.Set(rp.USBCTRL_REGS_SIE_STATUS_BUS_RESET)
+		initEndpoint(0, usb_ENDPOINT_TYPE_CONTROL)
+	}
+}
+
+func initEndpoint(ep, config uint32) {
+	val := uint32(0x80000000) | uint32(0x20000000)
```

Comment:
It would be great to have some const definitions for `0x80000000`, `0x20000000` and the other "magic numbers" here.

Reply:
Added const definition.